### PR TITLE
Create a generic number ↔ name and let EVP use it

### PR DESCRIPTION
**Pull request by levitte**
This creates a generic, per library context number ↔ name map, and has the EVP fetcher use it.

##### Checklist

- [x] documentation is added or updated

### CODE REVIEW

**Comment by levitte** [comment]
Arguably, this namemap could be used by the property code...  I don't quite understand why `crypto/property/property_string.c` currently holds two databases, when they are really just string ↔ identity maps and those strings don't have any other inherent per se.

**Comment by paulidale** [comment]
@levitte the reason that `crypto/property/property_string.c` holds two (name, value) databases is to distinguish between the _names_ and the _values_.   There are strict rules about what forms _names_ can take and these are enforced.  There are far laxer restrictions on _values_.  Keeping them separate seemed expedient from the point of view of validity checking.

I'm not going to object to a merging of the two, so long as the semantics don't also change.

File: crypto/core_namemap.c [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+    if (ossl_namemap_number(libctx, name) != 0)
+        return 1;                /* Pretend success */
+
+    if ((entry = OPENSSL_zalloc(sizeof(*entry) + strlen(name))) == NULL)
+        goto err;
+
+    strcpy(entry->body, name);
+    entry->name = entry->body;
+    entry->number = sk_NAMEMAP_push(store->numname, entry);
```
Comment (mattcaswell) [inline crypto/core_namemap.c]: Doesn't this need to be locked?

Comment (levitte) [reply crypto/core_namemap.c]: Yes, and I will add that. I was very short on time when I put together this PR...

File: crypto/core_namemap.c [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+#include <openssl/safestack.h>
+
+typedef struct {
+    int number;
+    const char *name;
+    char body[1];        /* Sized appropriately to contain the name */
+} NAMEMAP;
+
+DEFINE_LHASH_OF(NAMEMAP);
+DEFINE_STACK_OF(NAMEMAP)
```
Comment (paulidale) [inline crypto/core_namemap.c]: Sparse array instead here?
This kind of lookup is why it was added.

**Comment by paulidale** [comment]
There is a slight difference between the properties name ↔ value maps and this.  The property maps are known to be indexed sequentially (from zero) which makes a stack of names efficient, here the index doesn't have that property which makes a sparse array the better choice.

By switching the property implementation over, there will be a (small?) space cost but I don't think there will be a noticeable time cost.  I believe the simplification to a single name ↔ value map across the library is a beneficial one that outweighs associated the costs.

**Comment by paulidale** [comment]
Merging the two name ↔ value maps in the property code into a single map will allow properties like: "fips=foo, foo=bar" (where _foo_ is **not** a property name) to be specified and parsed without error.  I.e. the reservation of names without a period can be circumvented.

File: crypto/core_namemap.c (hunk 2 above)
Comment (levitte) [reply crypto/core_namemap.c]: Sure...

**Comment by levitte** [comment]
> There is a slight difference between the properties name ↔ value maps and this. The property maps are known to be indexed sequentially (from zero) which makes a stack of names efficient, here the index doesn't have that property which makes a sparse array the better choice.

Uhmmmm, this namemap is indexed monotonically as well, but from one (zero indicates an "undefined" string).  A string gets added by pushing it in the stack...

File: crypto/core_namemap.c (hunk 2 above)
Comment (levitte) [reply crypto/core_namemap.c]: Your discussion comments make me wonder, though, if you have the right idea about the code implemented here...

**Comment by levitte** [comment]
There's one bit that I find highly disturbing, and it's that this namemap is per library context...  That would mean duplicating it unnecessarily if we're handling several library contexts within a unit.  It's just a string to number cache, essentially, so there's no reason to make the numbers different per context.

I *could* force it to end up in the default library context, but that would mean that there must be one, and the FIPS module currently forbids that...  I might have an idea or two on that, btw, but that's the matter of another PR.

File: crypto/core_namemap.c (hunk 2 above)
Comment (paulidale) [reply crypto/core_namemap.c]: I'm wrong :(

**Comment by paulidale** [comment]
I saw the use of `sk_NAMEMAP_value` and thought that it searched for some unknown reason.

Best to stick with the stack and it can directly replace the two property mappings.

**Comment by paulidale** [comment]
Could we have this per context and default to the "default" context if none is directly set?

**Comment by paulidale** [comment]
Properties index from 1 as well.

**Comment by levitte** [comment]
>  Could we have this per context and default to the "default" context if none is directly set?

We can.  Also, this implementation only allows one namemap (per library context), and what you said about keeping property names and values separate makes me think that we need a mechanism to have more than one.  This is perfectly viable, but needs an extension of the API.

**Comment by levitte** [comment]
I re-implemented it to make it possible to create free floating name maps.

**Comment by levitte** [comment]
Ooooh, this found me an interesting bug...  we weren't doing things quite right when there is no pre-existing nid for a name...

**Comment by levitte** [comment]
This isn't in draft any more

File: crypto/evp/evp_fetch.c [hunk 3]
```diff
@@ -112,30 +113,13 @@ static void *construct_method(const char *algorithm_name,
...
-        tmpobj.flags = ASN1_OBJECT_FLAG_DYNAMIC;
-        tmpobj.length = 0;
-        tmpobj.data = NULL;
-
-        nid = OBJ_add_object(&tmpobj);
-    }
+    int nid;
 
-    if (nid == NID_undef)
+    if ((nid = ossl_namemap_number(methdata->libctx, algorithm_name)) == 0
```
Comment (slontis) [inline crypto/evp/evp_fetch.c]: The namemap_new seems to call ossl_namemap_number() and early out if its already mapped - so you may not need the first line.

Comment (levitte) [reply crypto/evp/evp_fetch.c]: Things have changed since this comment...  care to take a refreshed look?

File: crypto/core_namemap.c [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+} NAMEMAP_ENTRY;
+
+DEFINE_LHASH_OF(NAMEMAP_ENTRY);
+DEFINE_STACK_OF(NAMEMAP_ENTRY)
+
+/* The namemap, which provides for bidirectional indexing */
+
+struct ossl_namemap_st {
+    /* Flags */
+    unsigned int floating:1; /* If 0, it's stored in a library context */
```
Comment (slontis) [inline crypto/core_namemap.c]: Not sure about the name - maybe something alloc related would be better.

File: crypto/core_namemap.c [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+ *
+ * This isn't currently possible because of FIPS module constraints, so
+ * we currently disable the code that would allow it.
+ */
+
+const char *ossl_namemap_name(OSSL_NAMEMAP *namemap, int number)
+{
+    NAMEMAP_ENTRY *entry;
+
+#if 0                            /* TODO(3.0) */
```
Comment (slontis) [inline crypto/core_namemap.c]: Is this a #ifndef FIPS_MODE

File: crypto/evp/evp_fetch.c [hunk 6]
```diff
@@ -146,21 +135,22 @@ static void destruct_method(void *method, void *data)
 }
 
 void *evp_generic_fetch(OPENSSL_CTX *libctx, int operation_id,
-                        const char *algorithm, const char *properties,
```
Comment (slontis) [inline crypto/evp/evp_fetch.c]: why was this changed to name?

File: doc/internal/man3/ossl_namemap_new.pod [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+ #include "internal/cryptlib.h"
+
+ OSSL_NAMEMAP *ossl_namemap_stored(OPENSSL_CTX *libctx);
+
+ OSSL_NAMEMAP *ossl_namemap_new(void);
+ void ossl_namemap_free(OSSL_NAMEMAP *namemap);
+
+ int ossl_namemap_add(OSSL_NAMEMAP *namemap, const char *name);
+ const char *ossl_namemap_name(OSSL_NAMEMAP *namemap, int number);
```
Comment (slontis) [inline doc/internal/man3/ossl_namemap_new.pod]: const OSSL_NAMEMAP ?

File: doc/internal/man3/ossl_namemap_new.pod [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+ int ossl_namemap_number(OSSL_NAMEMAP *namemap, const char *name);
+
+=head1 DESCRIPTION
+
+A B<OSSL_NAMEMAP> is a simple number E<lt>-E<gt> name map, which can
+be used to give arbitrary any name (any string) a unique dynamic
```
Comment (slontis) [inline doc/internal/man3/ossl_namemap_new.pod]: any arbitrary

File: doc/internal/man3/ossl_namemap_new.pod [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+identity that is valid throughout the lifetime of the associated
+library context.
+
+ossl_namemap_new() and ossl_namemap_free() construct and destruct a
+new B<OSSL_NAMEMAP>.
+This is suitable to use when the B<OSSL_NAMEMAP> is embedded in other
+structures, or should be independent for any reason.
+
+ossl_namemap_stored() finds or a auto-creates the default namemap in
```
Comment (slontis) [inline doc/internal/man3/ossl_namemap_new.pod]: remove 'a'

File: include/internal/namemap.h [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+OSSL_NAMEMAP *ossl_namemap_stored(OPENSSL_CTX *libctx);
+
+OSSL_NAMEMAP *ossl_namemap_new(void);
+void ossl_namemap_free(OSSL_NAMEMAP *namemap);
+
+int ossl_namemap_add(OSSL_NAMEMAP *namemap, const char *name);
+const char *ossl_namemap_name(OSSL_NAMEMAP *namemap, int number);
+int ossl_namemap_number(OSSL_NAMEMAP *namemap, const char *name);
+
+
```
Comment (slontis) [inline include/internal/namemap.h]: nit space

File: crypto/core_namemap.c (hunk 4 above)
Comment (levitte) [reply crypto/core_namemap.c]: Well, it *is* allocated either way.  "stored"?

File: crypto/core_namemap.c (hunk 5 above)
Comment (levitte) [reply crypto/core_namemap.c]: No.  But FIPS module related changes I have in my mind will hopefully make this possible, and enabled at all times, unconditionally...  if it works out like I imagine (there will be another PR on this)

File: crypto/evp/evp_fetch.c (hunk 6 above)
Comment (levitte) [reply crypto/evp/evp_fetch.c]: For consistency.  We have "name", "algorithm_name" and "algorithm".  Since this is EVP, we *know* they are names of algorithms, and this parameter *is* the algorithm name, so...

File: doc/internal/man3/ossl_namemap_new.pod (hunk 7 above)
Comment (levitte) [reply doc/internal/man3/ossl_namemap_new.pod]: Good point

File: crypto/core_namemap.c (hunk 4 above)
Comment (slontis) [reply crypto/core_namemap.c]: yep sounds better

File: doc/internal/man3/ossl_namemap_new.pod (hunk 8 above)
Comment (levitte) [reply doc/internal/man3/ossl_namemap_new.pod]: Okie

File: doc/internal/man3/ossl_namemap_new.pod (hunk 7 above)
Comment (slontis) [reply doc/internal/man3/ossl_namemap_new.pod]: and then next one too

File: doc/internal/man3/ossl_namemap_new.pod (hunk 9 above)
Comment (levitte) [reply doc/internal/man3/ossl_namemap_new.pod]: Sure

File: include/internal/namemap.h (hunk 10 above)
Comment (levitte) [reply include/internal/namemap.h]: Good point

File: crypto/core_namemap.c [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+
+/* OPENSSL_CTX_METHOD functions for a namemap stored in a library context */
+
+static void *stored_namemap_new(OPENSSL_CTX *libctx)
+{
+    OSSL_NAMEMAP *namemap = ossl_namemap_new();
+
+    if (namemap != NULL) {
+        namemap->stored = 1;
+    }
```
Comment (paulidale) [inline crypto/core_namemap.c]: Lose the {}s ?

File: crypto/core_namemap.c [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+
+    return namemap;
+}
+
+static void stored_namemap_free(void *vnamemap)
+{
+    OSSL_NAMEMAP *namemap = vnamemap;
+
+    /* Pretend it's stored, or ossl_namemap_free() will do nothing */
```
Comment (paulidale) [inline crypto/core_namemap.c]: Is this meant to say _not_ stored?

File: crypto/core_namemap.c [hunk 13]
```diff
@@ -0,0 +1,217 @@
...
+    lh_NAMEMAP_ENTRY_free(namemap->namenum);
+
+    sk_NAMEMAP_ENTRY_pop_free(namemap->numname, namemap_free);
+
+    CRYPTO_THREAD_lock_free(namemap->lock);
+    OPENSSL_free(namemap);
+}
+
+/*
+ * TODO(3.0) Make NULL to signify the default namemap, found in the default
```
Comment (paulidale) [inline crypto/core_namemap.c]: I wonder if this is the best place to convert NULL to default?

- If the property string storage is modified to use this data structure, it would be bad.
- It would be okay using the property string data structure here, I think.
- The third obvious choice is to abstract the data structure and share it between both places.

File: crypto/core_namemap.c [hunk 14]
```diff
@@ -0,0 +1,217 @@
...
+OSSL_NAMEMAP *ossl_namemap_new(void)
+{
+    OSSL_NAMEMAP *namemap;
+
+    if ((namemap = OPENSSL_zalloc(sizeof(*namemap))) != NULL
+        && (namemap->lock = CRYPTO_THREAD_lock_new()) != NULL
+        && (namemap->numname = sk_NAMEMAP_ENTRY_new_null()) != NULL
+        && (namemap->namenum =
+            lh_NAMEMAP_ENTRY_new(namemap_hash, namemap_cmp)) != NULL) {
+        namemap->stored = 0;
```
Comment (mattcaswell) [inline crypto/core_namemap.c]: Not strictly necessary due to the zalloc

File: crypto/evp/evp_fetch.c [hunk 15]
```diff
@@ -68,16 +69,23 @@ static OSSL_METHOD_STORE *get_default_method_store(OPENSSL_CTX *libctx)
...
+    OSSL_NAMEMAP *namemap;
+    int nid;
 
     if (store == NULL
         && (store = get_default_method_store(libctx)) == NULL)
         return NULL;
 
-    (void)ossl_method_store_fetch(store, methdata->nid, propquery, &method);
+    if ((namemap = ossl_namemap_stored(libctx)) == NULL
+        || (nid = ossl_namemap_add(namemap, name)) == 0)
```
Comment (mattcaswell) [inline crypto/evp/evp_fetch.c]: While this is a nid in the sense that nid == numeric identifier, it is a different type of nid to where that term is used everywhere else in the code. I think that's confusing so a different name would be a good idea.

File: crypto/evp/evp_fetch.c [hunk 16]
```diff
@@ -107,35 +117,14 @@ static int put_method_in_store(OPENSSL_CTX *libctx, void *store,
...
-        ASN1_OBJECT tmpobj;
-
-        /* This is the same as OBJ_create() but without requiring a OID */
-        tmpobj.nid = OBJ_new_nid(1);
-        tmpobj.sn = tmpobj.ln = methdata->name;
-        tmpobj.flags = ASN1_OBJECT_FLAG_DYNAMIC;
-        tmpobj.length = 0;
-        tmpobj.data = NULL;
+    /* TODO(3.0) get rid of the need for legacy NIDs */
+    int legacy_nid = OBJ_sn2nid(name);
```
Comment (mattcaswell) [inline crypto/evp/evp_fetch.c]: Why is this needed? OBJ_sn2nid uses ASN1_OBJECT and I thought removing that dependency that was part of the point of this PR?

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: crypto/evp/evp_fetch.c (hunk 16 above)
Comment (levitte) [reply crypto/evp/evp_fetch.c]: The legacy NID needs to be preserved because of things like this:

https://github.com/openssl/openssl/blob/9acbe07d2300d34a7ea846d9756f33b4595e32fb/crypto/evp/evp_lib.c#L25-L37

File: crypto/evp/evp_fetch.c (hunk 15 above)
Comment (levitte) [reply crypto/evp/evp_fetch.c]: I can certainly go with `id` if that's preferable.

File: crypto/core_namemap.c (hunk 11 above)
Comment (levitte) [reply crypto/core_namemap.c]: Done

File: crypto/core_namemap.c (hunk 12 above)
Comment (levitte) [reply crypto/core_namemap.c]: Right.  Fixed

File: crypto/core_namemap.c (hunk 13 above)
Comment (levitte) [reply crypto/core_namemap.c]: Since the property strings need to be distinguished from all other strings, for reasons you've already explained, I would say that you won't use NULL, but rather allocate your own namemaps as part of this structure:

https://github.com/openssl/openssl/blob/9acbe07d2300d34a7ea846d9756f33b4595e32fb/crypto/property/property_string.c#L37-L42

In other words, the structure could be remade like this:

``` C
typedef struct {
    OSSL_NAMEMAP *prop_names;
    OSSL_NAMEMAP *prop_values;
} PROPERTY_STRING_DATA;
```

... and the property code would only keep the higher end strings functions.

File: crypto/core_namemap.c (hunk 14 above)
Comment (levitte) [reply crypto/core_namemap.c]: Removed

File: crypto/evp/evp_fetch.c (hunk 16 above)
Comment (levitte) [reply crypto/evp/evp_fetch.c]: Of course, I answered a bit too quickly, and only your first question.  I'll take a closer look at what I just quoted.

File: crypto/evp/evp_fetch.c [hunk 17]
```diff
@@ -107,35 +117,23 @@ static int put_method_in_store(OPENSSL_CTX *libctx, void *store,
...
+
+#ifndef FIPS_MODE
+    /* TODO(3.x) get rid of the need for legacy NIDs */
+    /*
+     * FIPS module note: since internal fetches will be entirely provider
+     * based, we know that none of its code depends on legacy NIDs or any
+     * functionality that use them.
+     */
+    legacy_nid = OBJ_sn2nid(name);
+#endif
```
Comment (levitte) [inline crypto/evp/evp_fetch.c]: So here's a point that may make this PR controversial.  I would like to propose that *within the FIPS module*, the "legacy" NIDs are (or should be) irrelevant.

In the particular case I identified where "legacy" NIDs are needed is all about ASN.1 parameters, something we won't have to deal with within the FIPS module as far as I understand.  I'll spend some time identifying others...

Comment (levitte) [reply crypto/evp/evp_fetch.c]: ... or perhaps not so controversial after all...  I just looked in #8728 and noticed that the "offending" code is excluded from the FIPS module, so...

File: crypto/evp/evp_fetch.c (hunk 16 above)
Comment (mattcaswell) [reply crypto/evp/evp_fetch.c]: The above code only needs to work for implicit lookup ciphers so shouldn't impact the evp_fetch code.

File: crypto/evp/evp_fetch.c (hunk 15 above)
Comment (mattcaswell) [reply crypto/evp/evp_fetch.c]: Yes, I think that would be less confusing.

Comment (levitte) [reply crypto/evp/evp_fetch.c]: Okie

File: crypto/core_namemap.c (hunk 5 above)
Comment (mattcaswell) [reply crypto/core_namemap.c]: I'd prefer not to see this "if 0" code here. I think its better to add it in whatever future PR enables this functionality. It doesn't really add anything here since I can't see the big picture of how you intend to use this capability.

File: doc/internal/man3/ossl_namemap_new.pod [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+ void ossl_namemap_free(OSSL_NAMEMAP *namemap);
+
+ int ossl_namemap_add(OSSL_NAMEMAP *namemap, const char *name);
+ const char *ossl_namemap_name(const OSSL_NAMEMAP *namemap, int number);
+ int ossl_namemap_number(const OSSL_NAMEMAP *namemap, const char *name);
+
+=head1 DESCRIPTION
+
+A B<OSSL_NAMEMAP> is a simple number E<lt>-E<gt> name map, which can
+be used to give arbitrary any arbitrary name (any string) a unique
```
Comment (mattcaswell) [inline doc/internal/man3/ossl_namemap_new.pod]: arbitrary any arbitrary?

File: doc/internal/man3/ossl_namemap_new.pod [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+
+ossl_namemap_number() finds the number corresponding to the given
+name.
+
+=head1 NOTES
+
+It's possible that ossl_namemap_add(), ossl_namemap_name() and
+ossl_namemap_number() will change to accept NULL as a B<OSSL_NAMEMAP>
+to use one that is auto-created and associated with the default
+library context.
```
Comment (mattcaswell) [inline doc/internal/man3/ossl_namemap_new.pod]: I'd delete the above. Either it does get added in the future - in which case we document it then. Or it doesn't, and this note becomes dead.

File: util/find-doc-nits [hunk 20]
```diff
@@ -198,8 +199,6 @@ sub check()
         if $contents =~ /=head\d\s\s+/;
     print "$id period in NAME section\n"
         if $contents =~ /=head1 NAME.*\.\n.*=head1 SYNOPSIS/ms;
-    print "$id POD markup in NAME section\n"
-        if $contents =~ /=head1 NAME.*[<>].*=head1 SYNOPSIS/ms;
```
Comment (mattcaswell) [inline util/find-doc-nits]: Not sure I understand the purpose of this find-doc-nits change.

**Review by mattcaswell — CHANGES_REQUESTED** [review]
Also the nid->id change is still outstanding

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: crypto/core_namemap.c [hunk 21]
```diff
@@ -0,0 +1,216 @@
...
+    lh_NAMEMAP_ENTRY_free(namemap->namenum);
+
+    sk_NAMEMAP_ENTRY_pop_free(namemap->numname, namemap_free);
+
+    CRYPTO_THREAD_lock_free(namemap->lock);
+    OPENSSL_free(namemap);
+}
+
+/*
+ * TODO(3.0) Make NULL to signify the default namemap, found in the default
```
[the remark this answers is not in the document]
Comment (mattcaswell) [reply crypto/core_namemap.c]: Right....that doesn't really answer the question. We're not using this capability as far as I can tell, so its not clear why its useful.

[the remark this answers is not in the document]
Comment (levitte) [reply crypto/core_namemap.c]: Oh wait, yeahok, I see what you mean.  Fix coming up

[the remark this answers is not in the document]
Comment (levitte) [reply crypto/core_namemap.c]: This is a generic namenum map.  With the current implementation, every library context will have its own default one, so for the string "foo", there will be as many entries, when the mapped numbers could just as well be exactly the same for all of them.

**Comment by levitte** [comment]
Ping?

(btw, in a conversation yesterday, it was suggested that the strings could be bitstrings, i.e. binary data rather than just text...  I don't mind the idea, but I also think that could be future development of this map functionality, it doesn't have to happen in *this* PR)

**Comment by levitte** [comment]
Doing the close+open thing to kick Appveyor

**Review by paulidale — APPROVED** [review]

**Comment by levitte** [comment]
Merged.

0211740fcc47a954be19cceb65fb57a6f7deb797 EVP_FETCH: remove the need to transport the legacy NID through construction
1f79ddf5049ff53ad8a7cbab76e62d02d9ac099f util/find-doc-nits: Fine tune detection of POD markup in NAME section
2e49c05472ab76cee4e30c2eaa4fa576b9ae92c6 EVP_FETCH: deal with names without pre-defined NIDs
baff732dc654f8aab37953f6036399989c765604 Make the generic EVP fetching mechanism use the namenum map
f2182a4e6fef098014117eb4a4f35ca84cd2cf8e Create internal number<->name mapping API
